#50102 Task 3: Until, backed by retry watcher

tnozicka opened the pull request:
**What this PR does / why we need it**:
This is a split off from https://github.com/kubernetes/kubernetes/pull/50102 to go in smaller pieces.

Introduces Until based on RetryWatcher. It can survive closed watches if last read ResourceVersion is still present in etcd.

Fixes: https://github.com/kubernetes/kubernetes/issues/31345

**Requires**:
 - [x] https://github.com/kubernetes/kubernetes/pull/66906
 - [x] https://github.com/kubernetes/kubernetes/pull/66971


/hold

**Release note**:
```release-note
watch.Until now works for long durations.
```

/priority important-soon
/kind bug
(bug after the main PR which is this split from)

spiffxp commented:
ping, curious if any progress is being planned for this release cycle

spiffxp commented:
/skip
trying to clear the Submit Queue status

tnozicka committed:
Add Until based on RetryWatcher

tnozicka commented:
/milestone v1.14

spiffxp commented:
/area deflake
#50102 does a good job explaining why this has the potential to address a number of flakes

tnozicka commented:
I have just removed the dependency on https://github.com/kubernetes/kubernetes/pull/72566 which hasn't been accepted and rebased so this should be ready. I'll do one pass to sanity check this for bugs tomorrow but this is ready to be looked at.

tnozicka commented:
/refresh

tnozicka commented:
hm, CI seems stuck on `Expected — Waiting for status to be reported`
/test all

tnozicka commented:
/hold cancel

now ready for review

@wojtek-t @liggitt @kubernetes/sig-api-machinery-pr-reviews

tnozicka commented:
/retest

tnozicka commented:
/retest

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go:
```diff
@@ -0,0 +1,509 @@
+}
+
+func AtoiOrDie(s string) int {
+	i, err := strconv.Atoi(s)
```
clients should never assume the resource version is an int... the only valid test of a resource version is an equality check

from https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata:

> resourceVersion: a string that identifies the internal version of this object that can be used by clients to determine when objects have changed. This value MUST be treated as opaque by clients and passed unmodified back to the server. Clients should not assume that the resource version has meaning across namespaces, different kinds of resources, or different servers.

liggitt replied:
oh... just realized this is in a test file... in that case, this is acting as the server, so interpreting it is ok

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go:
```diff
@@ -0,0 +1,509 @@
+	}
+}
+
+func AtoiOrDie(s string) int {
```
make this private and add a comment that this is test-only code that can only interpret resourceversion because it is simulating the server

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+
+// receive reads the result from a watcher, restarting it if necessary.
+func (rw *RetryWatcher) receive() {
+	defer close(rw.doneChan)
```
This is closing the same channel that is closed by RetryWatcher#Stop. Won't that cause a panic?

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+	if err != nil {
+		klog.Errorf("retryWatcher: watcherFunc failed: %v", err)
+		// Retry
+		return false, 0
```
won't this hot-loop retry on network errors?

tnozicka replied:
It will, but it will always check for the stop channel so the timeout (context) from Until cancels it. This used to hard fail, but in this last iteration I have decided for approach more similar to informers. I am  open to other views on how we should treat errors here, but my latest thinking is proceed at all cost, unless you know for sure you can't - network errors, permissions it all can be transilient

tnozicka replied:
added artificial waits for hard failures

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+
+// doReceive returns true when it is done, false otherwise.
+// If it is not done the second return value holds the time to wait before calling it again.
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
```
if `rw.lastResourceVersion` is `""` or `"0"`, send an error and `return true, 0`

tnozicka replied:
done in the constructor

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+			return true, 0
+		case event, ok := <-ch:
+			if !ok {
+				klog.Warningf("RetryWatcher - getting event failed! Re-creating the watcher. Last RV: %s", rw.lastResourceVersion)
```
isn't the channel closing a normal occurrence? warning level logging seems way too high... I'd have thought v4 or so

tnozicka replied:
you are right

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+					}
+
+					// Retry after delay
+					return false, time.Duration(status.Details.RetryAfterSeconds) * time.Second
```
is there a reason to limit this to StatusGatewayTimeout, and not to use `api/error#SuggestsClientDelay(err)` to check this?

tnozicka replied:
thx, I wasn't aware we had a function for it, will fix it

tnozicka replied:
I've made it common for all the failures but didn't use `SuggestsClientDelay` in the end as it requires  different type and creating another object embedding the type just to artificially get it din't feel right

liggitt replied:
suggest extracting a helper method to check for delay directly on a status object in the api errors package like this:

```go
// SuggestsClientDelay returns true if this error suggests a client delay as well as the
// suggested seconds to wait, or false if the error does not imply a wait. It does not
// address whether the error *should* be retried, since some errors (like a 3xx) may
// request delay without retry.
func SuggestsClientDelay(err error) (int, bool) {
	if t, ok := err.(APIStatus); ok {
		return StatusSuggestsClientDelay(t.Status())
	}
	return 0, false
}

func StatusSuggestsClientDelay(status metav1.Status) (int, bool) {
	if status.Details == nil {
		return 0, false
	}
	if status.Reason == metav1.StatusReasonServerTimeout {
		// this StatusReason explicitly requests the caller to delay the action
		return int(status.Details.RetryAfterSeconds), true
	}
	if status.Details.RetryAfterSeconds > 0 {
		// If the client requests that we retry after a certain number of seconds
		return int(status.Details.RetryAfterSeconds), true
	}
	return 0, false
}
```

we really don't want to treat status codes and retryafter delays inconsistently

tnozicka replied:
hm, given we'd just ignore the bool value, that whole function comes down to: give me whatever value is in `RetryAfterSeconds`. 

If that's 0, retry immediately, wait the amount if not.

what Am I missing?

(I guess I can just extract the function anyways to move it forward, just don't see the value.)

tnozicka replied:
Also none of the usages of that function actually reads the int value (I wondered why we don't return time.Duration there) so it seems exclusively used to determine retry purposes given you can't infer it  being set from the int value.

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+			klog.V(4).Info("Stopping RetryWatcher.")
+			return
+		case <-time.After(retryAfter):
+			done, retryAfter = rw.doReceive()
```
other loops that re-establish watches do so at a 1 second interval. how are we avoiding hot-looping here?

tnozicka replied:
are those (1s) mostly controller loops? my thinking is - watch getting closed is a normal occurrence and we should reestablish it as fast as we can, as this should behave the same way as a singe watch and reestablishing it shouldn't be observed nor cause artificial delays, at least for the first time - maybe we need to limit consequent failures or we are good with leaving this on client rate-limiter?

tnozicka replied:
delays are now added in the inner function

liggitt replied:
I still see several places it does `return false, 0`

liggitt commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -95,6 +95,17 @@ func UntilWithoutRetry(ctx context.Context, watcher watch.Interface, conditions
+// It guarantees you to see all events and in the order they happened.
+// Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
+// The most frequent usage would be a test where you want to verify exact order of events.
+func Until(ctx context.Context, initialResourceVersion string, watcherFunc WatcherFunc, conditions ...ConditionFunc) (*watch.Event, error) {
```
verify `initialResourceVersion` is not "" or "0" (return error if it is)?

tnozicka replied:
could you help me understand why those are invalid values in this context?

liggitt replied:
The synthetic ADDED watch events you get back for existing objects when starting a watch without specifying a resourceVersion are not ordered by resourceVersion, so restarting a broken watch request based on the resourceVersion of one of the received events (as the RetryWatcher does) is not safe.

That could be improved in the future, and if it is, the restriction here could be lifted.

tnozicka replied:
watch cache?
https://github.com/kubernetes/kubernetes/blob/2cbb16bc8dd456c5db72c1667926abdbc87c32c7/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L434-L461
listing directly from the store and breaking the event order sounds like a bug to me. I'll send a PR to fix it and we don't have to hack it here, also I want this to work from `""`.

liggitt replied:
It is not watch cache specific. It happens on resources that do not use watch cache as well

tnozicka replied:
hm, if we don't use watch cache there, we are reading etcd, which I suppose has index on RV so reading it from the index should be preferable and ordered. (In the worst case the watch start 0 is basically a list there which we could order before sending it.)

What other places of breaking the invariant except watch cache init are you aware? are those just regular etcd reads? I think we need to fix those. The invariant of a watch call is (should be) that events are deliver in the order they happened (ordered by RV).

liggitt replied:
> which I suppose has index on RV

That gets compacted after 5 minutes. List from etcd is not ordered by RV.

> In the worst case the watch start 0 is basically a list there which we could order before sending it

that is potentially a sort of a list of tens or hundreds of thousands of items

wojtek-t replied:
https://github.com/kubernetes/kubernetes/issues/13969

And I agree with Jordan it's not easily fixable.

liggitt replied:
requiring a specific resource version to start with here doesn't seem unreasonable to me, especially since this method's purpose is to ensure observation of every watch event, a guarantee you don't get when starting from ""/"0"

tnozicka replied:
I consider correctness our decision point. 

If my watch call get interrupted in the initial list phase, which I have no way to detect because it's just a first part of that continuous stream - how do I restart the watch? if I start from last RV received and the stream is not ordered I'll receive some events twice or skip some.

tnozicka replied:
I admit this may be slipping to how a single WATCH call is broken, not this handling.

I don't think we need v2 to fix that, although it would avoid the issue we are having but v2 premise is API change which gets hard.

I am pretty new to this part of the machinery, but looks like the fix so far requires a ´few lines in a watch cache and listing from etcd using WithSort()

liggitt replied:
> If my watch call get interrupted in the initial list phase, which I have no way to detect because it's just a first part of that continuous stream - how do I restart the watch? if I start from last RV received and the stream is not ordered I'll receive some events twice or skip some.

Yes, that is currently the issue with watches started from "" or "0".

> I admit this may be slipping to how a single WATCH call is broken, not this handling.

I'm open to investigating improving that, but in the meantime, this should prevent usage we know is flawed.

> I am pretty new to this part of the machinery, but looks like the fix so far requires a ´few lines in a watch cache and listing from etcd using WithSort()

We can investigate both of those, to see if we can do them performantly.

liggitt replied:
sorting (either in the apiserver from the watch cache, or in etcd for direct lists) would be in-memory... neither maintains an ordered index by revision:

https://github.com/kubernetes/kubernetes/blob/01e7b3040a1f47c613e6db85f9bccd727f5ef89c/staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go#L441-L460

https://github.com/etcd-io/etcd/blob/24fc2a983ae39b4fe09cef3db434c812fdc97cff/etcdserver/apply.go#L313-L323

liggitt replied:
opened exploratory PR with those changes at https://github.com/kubernetes/kubernetes/pull/74020

tnozicka replied:
cool, I've wrote up the issue here https://github.com/kubernetes/kubernetes/issues/74022 to direct the discussion and started writing some unit tests to prove the behavior locally, the PR is what I had in mind, thx for looking into it

tnozicka replied:
Until we sort out that approach in https://github.com/kubernetes/kubernetes/issues/74022 I've restricted using those values

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,232 @@
+func (rw *RetryWatcher) doReceive() (bool, time.Duration) {
+	watcher, err := rw.watcherFunc(rw.lastResourceVersion)
+	if err != nil {
+		klog.Errorf("retryWatcher: watcherFunc failed: %v", err)
```
wondering if this needs to do similar things to https://github.com/kubernetes/kubernetes/blob/f96849981203ea506ea8fa3f88c173d36cf688aa/staging/src/k8s.io/client-go/tools/cache/reflector.go#L276-L299 with respect to avoiding unnecessarily loud logging

tnozicka replied:
it is fairly unlikely since it just the beginning of the watch call, not the place where we wait, but now it's covered as well

tnozicka committed:
Upgrade ListWatchUntil

tnozicka committed:
Update Bazel

anarchistHH1983 commented:
/retest

On Mon, Feb 18, 2019, 1:54 PM Kubernetes Prow Robot <
notifications@github.com wrote:

> @tnozicka <https://github.com/tnozicka>: The following test *failed*, say
> /retest to rerun them all:
> Test name Commit Details Rerun command
> pull-kubernetes-godeps c974742
> <https://github.com/kubernetes/kubernetes/commit/c974742d9828a170b415f1564244222c7b4c41ad>
> link
> <https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/67350/pull-kubernetes-godeps/2543> /test
> pull-kubernetes-godeps
>
> Full PR test history <https://gubernator.k8s.io/pr/67350>. Your PR
> dashboard <https://gubernator.k8s.io/pr/tnozicka>. Please help us cut
> down on flakes by linking to
> <https://git.k8s.io/community/contributors/devel/flaky-tests.md#filing-issues-for-flaky-tests>
> an open issue
> <https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open> when
> you hit one in your PR.
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/67350#issuecomment-464842498>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/Atif5fjIH9304Fxbfc90TGAkndu6DJ8Bks5vOvbzgaJpZM4V7Kg4>
> .
>

tnozicka commented:
@liggitt thanks for your insights, comments addressed, ptal

tnozicka commented:
just a minor update to extend `Until`'s description
```diff
diff --git a/staging/src/k8s.io/client-go/tools/watch/until.go b/staging/src/k8s.io/client-go/tools/watch/until.go
index 519618a42d..08c73a6d46 100644
--- a/staging/src/k8s.io/client-go/tools/watch/until.go
+++ b/staging/src/k8s.io/client-go/tools/watch/until.go
@@ -102,7 +102,9 @@ func UntilWithoutRetry(ctx context.Context, watcher watch.Interface, conditions
 // Until can deal with API timeouts and lost connections.
 // It guarantees you to see all events and in the order they happened.
 // Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
-// The most frequent usage would be a test where you want to verify exact order of events.
+// (See `UntilWith` sync if you'd prefer to recover from all the errors including RV too old by re-listing
+//  those items. In normal code you should care about being level driven so you'd not care about not seeing all the edges.)
+// The most frequent usage for Until would be a test where you want to verify exact order of events ("edges").
 func Until(ctx context.Context, initialResourceVersion string, watcherClient cache.Watcher, conditions ...ConditionFunc) (*watch.Event, error) {
 	w, err := NewRetryWatcher(initialResourceVersion, watcherClient)
 	if err != nil {
```

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,272 @@
+	case "", "0":
+		// TODO: revisit this if we ever get WATCH v2 where it means start "now"
+		//       without doing the synthetic list of objects at the beginning
+		return nil, fmt.Errorf("initial RV %q is not supported due to issues with underlying WATCH, see #74022", initialResourceVersion)
```
I'd avoid referencing issue numbers in error messages (including it in a comment here is fine)

tnozicka replied:
done

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,272 @@
+
+	case io.EOF:
+		// watch closed normally
+		return false, 0
```
anywhere this does `return false, 0`, we're still setting ourselves up for a hotloop, right?

tnozicka replied:
I think on the normal path, WATCH should be restarted immediately. 

What is a normal path could be the question. If you are unlucky to open a connection to apiserver that is just about to terminate, you likely want to start again by trying another watch as that wasn't error path you'd have to backoff on.

If your watch timeouts, you shouldn't sleep but try immediately.

I guess we could count consequent failures before getting to the processing phase but splitting failures into ones we want to retry immediately and ones we want to apply delay to seemed good enough for this use.

liggitt replied:
> If your watch timeouts, you shouldn't sleep but try immediately.

why? reflectors delay for a second in exactly that scenario, so they don't hotloop:

https://github.com/kubernetes/kubernetes/blob/8d6f20ea4342327bb778a80c70c47ec9ad47e182/staging/src/k8s.io/client-go/tools/cache/reflector.go#L113

https://github.com/kubernetes/kubernetes/blob/8d6f20ea4342327bb778a80c70c47ec9ad47e182/staging/src/k8s.io/client-go/tools/cache/reflector.go#L133-L137

I think retryAfter should be a minimum of 1 second.

tnozicka replied:
To me, the difference is that informers are caches and this is a form of direct api call.

Caches are ok with a delay, I don't expect delays to be injected into a real-time api call on the happy path.

tnozicka replied:
but I think there is a way to address both concerns, let me update the PR...

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,272 @@
+					_ = rw.send(event)
+					return true, 0
+
+				default:
```
a default switch case with a fallthrough to specific status codes reads really strangely... is there a clearer way to express this?

tnozicka replied:
moved it to a variable avoiding the fallthrough

liggitt commented on staging/src/k8s.io/client-go/tools/watch/until.go:
```diff
@@ -95,6 +95,25 @@ func UntilWithoutRetry(ctx context.Context, watcher watch.Interface, conditions
+// Until can deal with API timeouts and lost connections.
+// It guarantees you to see all events and in the order they happened.
+// Due to this guarantee there is no way it can deal with 'Resource version too old error'. It will fail in this case.
+// (See `UntilWith` sync if you'd prefer to recover from all the errors including RV too old by re-listing
```
should this be `UntilWithSync`?

tnozicka commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,282 @@
+
+	// We use non sliding until so we don't introduce delays on happy path when WATCH call
+	// timeouts or gets closed and we need to reestablish it while also avoiding hot loops.
+	wait.NonSlidingUntilWithContext(ctx, func(ctx context.Context) {
```
@liggitt this is addressing your hot loop concern without affecting the happy path.

In the end no matter what happens it won't restart sooner then in 1 second. But if a long running call (WATCH) fails e.g. after a minute it will restart immediately. If the restart fails, it won't get tried again sooner then in 1 sec.

liggitt replied:
looks good, thanks

tnozicka commented:
/test pull-kubernetes-bazel-test
(scheduler flake)

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,282 @@
+
+// receive reads the result from a watcher, restarting it if necessary.
+func (rw *RetryWatcher) receive() {
+	defer close(rw.doneChan)
```
don't we need to `defer close(rw.resultChan)` as well? doc says it will be closed when `Stop()` is called, and I don't see that happening (needs a test as well):

```
// Returns a chan which will receive all the events. If an error occurs
// or Stop() is called, this channel will be closed, in which case the
// watch should be completely cleaned up.
ResultChan() <-chan Event
```

tnozicka replied:
good catch, fixed

liggitt commented:
hotloop mitigation looks good, thanks. just noticed we're not closing ResultChan() on Stop(). lgtm once that's addressed

tnozicka commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go:
```diff
@@ -0,0 +1,591 @@
+		t.Error("Failed to close the watcher")
+	}
+
+	// RetryWatcher result channel should be closed
```
+ a test

liggitt commented on staging/src/k8s.io/client-go/tools/watch/retrywatcher.go:
```diff
@@ -0,0 +1,287 @@
+	// We need to wait for the receive goroutine to finish to avoid it writing to a closed channel
+	<-rw.Done()
+
+	close(rw.resultChan)
```
shouldn't this be done as a defer in `receive()`? then we wouldn't need to block on `<-rw.Done()` inside `Stop()`

tnozicka replied:
yeah, I shouldn't write code while in a meeting :)
fixed

tnozicka commented:
@liggitt thank you for reviewing this in detail!

tnozicka commented:
moving the ResultChan close has confused unit test, fixed:
```diff
diff --git a/staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go b/staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go
index 7e09fa953e..cd57e51c52 100644
--- a/staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go
+++ b/staging/src/k8s.io/client-go/tools/watch/retrywatcher_test.go
@@ -541,8 +541,10 @@ func TestRetryWatcher(t *testing.T) {
 			// but have to tolerate some delay. Given this is best effort detection we can use short duration.
 			// It also makes sure that for 0 events the watchFunc has time to be called.
 			select {
-			case event := <-watcher.ResultChan():
-				t.Error(spew.Errorf("Unexpected event received after reading all the expected ones: %#+v", event))
+			case event, ok := <-watcher.ResultChan():
+				if ok {
+					t.Error(spew.Errorf("Unexpected event received after reading all the expected ones: %#+v", event))
+				}
 			case <-time.After(10 * time.Millisecond):
 				break
 			}
```

liggitt commented:
/lgtm

nikhita commented:
Is this a flake? https://github.com/kubernetes/kubernetes/issues/56876#issuecomment-467864609

(can't reproduce locally)

nikhita commented:
cc @sttts @dims 
re above ^ question, because it is a publishing-bot failure

liggitt commented:
yes, looks like there are some races:

```sh
go test ./vendor/k8s.io/client-go/tools/watch -run TestRetryWatcher -count 1000 -race -v -failfast
```

produced at least these failures:
```
--- FAIL: TestRetryWatcher/survives_4_closed_watches_and_reads_4_items_for_nonconsecutive,_spread_RVs_and_skips_those_with_lower_or_equal_RV (0.03s)
    retrywatcher_test.go:554: expected 5 watcher starts, but it has started 4 times
```

```
--- FAIL: TestRetryWatcher/survives_1_closed_watch_and_reads_1_item (0.02s)
    retrywatcher_test.go:554: expected 2 watcher starts, but it has started 1 times
```

```
--- FAIL: TestRetryWatcher/survives_2_closed_watches_and_reads_2_items (0.04s)
    retrywatcher_test.go:554: expected 3 watcher starts, but it has started 2 times
```

tnozicka commented:
I'll fix send a fix shortly, it's a race in the unit tests when counting the watch calls (the last one is theoretically optional based on timing). Looks like I've used too small `-count` when testing it.

tnozicka commented:
fix is in https://github.com/kubernetes/kubernetes/pull/74663
